bwapi/BWAPI/Source/BWAPI/CommandTemp.h

```diff
@@ -12,712 +12,995 @@ namespace BWAPI
   template <class UnitImpl, class PlayerImpl>
   class CommandTemp
   {
-  public :
-    CommandTemp(const UnitCommand& command);
-    void execute(int frame);
-  protected:
-    virtual ~CommandTemp() {};
-  private :
+  public:
+    CommandTemp(const UnitCommand& command) : command(command) { }
+
+    void execute();
+    void execute(bool isCurrentFrame);
+
+    template<typename Buf>
+    void insertIntoCommandBuffer(Buf &buf) && {

```

Comment:
What is this syntax?

Reply:
Since the buffer is `vector<vector<Command>>`, and `CommandTemp<...>` is what makes up the type `Command`, this lets the caller use any `vector<vector<T>>` buffer, just to avoid any circular dependenices.

Reply:
Oh wait, this is why I made `Command` an alias instead of its own class. Guess I could change that to be based on the template parameters now then.

Reply:
You know what? While I made that edit I just realized that you meant the `&&` part. That means the function renders the object unusable, since there is a `std::move(*this)` inside of it when it moves itself into the buffer.

Reply:
See [here](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2439.htm) for the original proposal and [this](https://akrzemi1.wordpress.com/2014/06/02/ref-qualifiers/) for a explanation. This is how you make `*this` become an rvalue.

Reply:
You can see how this is called [here](https://github.com/bwapi/bwapi/pull/783/files#diff-d6f5e3e693825de7c06ae7858832499aL17)